### PR TITLE
Merge Inbox prove and process

### PR DESCRIPTION
**Pull request by nambrot**
This PR removes the notion of separate `prove` and `process` functions on the Inbox. It is motivated by the fact that `process` will need all the parameters of `prove` anyways for Sovereign Consensus, we already always called `proveAndProcess` in the first place.

### CODE REVIEW

**Review by tkporter — COMMENTED** [review]
looks good, just some small stuff

File: solidity/core/test/inbox.test.ts [hunk 1]
```diff
@@ -104,101 +105,47 @@ describe('Inbox', async () => {
...
   });
 
-  it('Rejects an already-proven message', async () => {
-    const testCase = merkleTestCases[0];
-    let { leaf, index, path } = testCase.proofs[0];
+  it('Rejects an already-processed message', async () => {
+    let { leaf, index, proof, root, message } = messageWithProof;
 
-    await inbox.setCheckpoint(testCase.expectedRoot, 1);
+    await inbox.setCheckpoint(root, 1);
```
Comment (asaj) [inline solidity/core/test/inbox.test.ts:121]: root, index + 1?

**Review by asaj — APPROVED** [review]
Approved for non-agent changes

File: solidity/core/contracts/Inbox.sol [hunk 2]
```diff
@@ -108,79 +106,54 @@ contract Inbox is IInbox, Version0, Common {
...
         bytes calldata _message,
         bytes32[32] calldata _proof,
-        uint256 _index
+        uint256 _index,
+        bytes calldata /* _sovereignData */
     ) external override {
-        require(prove(keccak256(_message), _proof, _index), "!prove");
-        process(_message);
+        // check re-entrancy guard
+        require(entered == 1, "!reentrant");
```
[the remark this answers is not in the document]
Comment (nambrot) [reply solidity/core/contracts/Inbox.sol:125]: I'm inclined to leave the semantics as is, as I just copied them?

File: solidity/core/contracts/Inbox.sol [hunk 3]
```diff
@@ -28,12 +28,10 @@ contract Inbox is IInbox, Version0, Common {
     // ============ Enums ============
 
     // Status of Message:
-    //   0 - None - message has not been proven or processed
-    //   1 - Proven - message inclusion proof has been validated
-    //   2 - Processed - message has been dispatched to recipient
+    //   0 - None - message has not been processed
```
[the remark this answers is not in the document]
Comment (nambrot) [reply solidity/core/contracts/Inbox.sol:31]: Probably, but I didn't want to make too far reaching changes, so I just kept this for now

File: solidity/core/contracts/Inbox.sol [hunk 4]
```diff
@@ -108,79 +106,54 @@ contract Inbox is IInbox, Version0, Common {
...
+     * @param _message Formatted message (refer to Common.sol Message library)
      */
-    function process(bytes calldata _message) public {
+    function _process(bytes calldata _message) internal {
         bytes29 _m = _message.ref(0);
         // ensure message was meant for this domain
         require(_m.destination() == localDomain, "!destination");
-        // ensure message has been proven
+
         bytes32 _messageHash = _m.keccak();
```
[the remark this answers is not in the document]
Comment (nambrot) [reply solidity/core/contracts/Inbox.sol]: I think it works fine to just pass the messageHash to the internal `_process` function? I agree that the separation is a big awkward, and that validation is split among both functions, but it's mainly to accomodate testing as you mentioned

File: rust/chains/abacus-ethereum/abis/Inbox.abi.json [hunk 5]
```diff
@@ -87,6 +87,19 @@
...
   },
+  {
+    "inputs": [
+      {
+        "internalType": "bytes",
+        "name": "_message",
+        "type": "bytes"
+      }
+    ],
+    "name": "_process",
```
[the remark this answers is not in the document]
Comment (nambrot) [reply rust/chains/abacus-ethereum/abis/Inbox.abi.json]: Yeah I must have copied it before I made it internal?

File: solidity/core/contracts/Inbox.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (tkporter) [reply solidity/core/contracts/Inbox.sol]: yeah works for me

File: solidity/core/test/inbox.test.ts (hunk 1 above)
Comment (nambrot) [reply solidity/core/test/inbox.test.ts:116]: Not really relevant to the test, right?

**Pull request merged**